### PR TITLE
Separate Firewaller from Environ interface

### PR DESCRIPTION
**Pull request by mikemccracken**
(WIP because I haven't tested this other than running unit tests yet)

Part of an effort to separate the Provider interface into minimal
interfaces, so that CAAS providers can be implemented simply.

This separates the Firewaller interface from Environ.

Removes unused implementations of the firewalling functions from the lxd and maas providers

Also results in separating InstanceFirewaller from Instance, because
those functions wrap the environ functions.

### CODE REVIEW

File: worker/firewaller/firewaller.go [hunk 1]
```diff
@@ -573,7 +573,13 @@ func (fw *Firewaller) reconcileInstances() error {
 			return err
 		}
 		machineId := machined.tag.Id()
-		initialRules, err := instances[0].IngressRules(machineId)
+
+		fwInstance, ok := instances[0].(instance.InstanceFirewaller)
+		if !ok {
+			return errors.Errorf("firewaller using unexpected instance type")
```
Comment (axw) [inline worker/firewaller/firewaller.go]: why error? shouldn't we just be ignoring the providers/instances that don't implement firewalling?

File: worker/firewaller/manifold.go [hunk 2]
```diff
@@ -86,6 +86,13 @@ func (cfg ManifoldConfig) start(context dependency.Context) (worker.Worker, erro
 	if err := context.Get(cfg.EnvironName, &environ); err != nil {
 		return nil, errors.Trace(err)
 	}
+
+	fwEnv, ok := environ.(environs.Firewaller)
+	if !ok {
+		logger.Infof("stopping firewaller (not used in this provider)")
+		return nil, dependency.ErrUninstall
```
Comment (axw) [inline worker/firewaller/manifold.go]: what about providers implementing instance firewalling?

File: instance/instance.go [hunk 3]
```diff
@@ -39,7 +39,9 @@ type Instance interface {
 	// Addresses returns a list of hostnames or ip addresses
 	// associated with the instance.
 	Addresses() ([]network.Address, error)
+}
 
+type InstanceFirewaller interface {
```
Comment (axw) [inline instance/instance.go]: please add a doc comment

File: worker/firewaller/firewaller.go [hunk 4]
```diff
@@ -871,17 +877,23 @@ func (fw *Firewaller) flushInstancePorts(machined *machineData, toOpen, toClose
 	if err != nil {
 		return err
 	}
+	fwInstance, ok := instances[0].(instance.InstanceFirewaller)
+	if !ok {
+		logger.Infof("flushInstancePorts called on an instance of type %T which doesn't support firewall.", instances[0])
```
Comment (axw) [inline worker/firewaller/firewaller.go]: Sorry, I'm an idiot. It didn't occur to me that this method would only be called if the relevant firewall mode were configured. So I think returning an error is sensible after all.

File: worker/firewaller/manifold.go [hunk 5]
```diff
@@ -86,10 +86,21 @@ func (cfg ManifoldConfig) start(context dependency.Context) (worker.Worker, erro
...
+	// nil value, as it won't be used.
+	fwEnv, fwEnvOK := environ.(environs.Firewaller)
+
 	mode := environ.Config().FirewallMode()
 	if mode == config.FwNone {
 		logger.Infof("stopping firewaller (not required)")
 		return nil, dependency.ErrUninstall
+	} else if mode == config.FwGlobal {
+		if !fwEnvOK {
+			logger.Infof("Firewall global mode set on provider with no support. stopping firewaller")
```
Comment (axw) [inline worker/firewaller/manifold.go]: IMO this should be an error, but I guess we have cases where we weren't rejecting FwGlobal, and have a dummy implementation? Oh well, I guess this is OK.

**Comment by mikemccracken** [comment]
$$merge$$

**Pull request merged**